### PR TITLE
Add WeightedAverageProfileParameter

### PR DESCRIPTION
**Pull request by Batch21**
This parameter uses the volume of given storage nodes to combine multiple parameters into a weighted profile.

### CODE REVIEW

File: pywr/parameters/_control_curves.pyx [hunk 1]
```diff
@@ -548,3 +549,78 @@ cdef class ControlCurveParameter(BaseControlCurveParameter):
...
+cdef class WeightedAverageProfileParameter(Parameter):
+    """Generates a weighted average daily profile from pairs of storage nodes and
+    profile parameters.
+
+    The contribution on each profile is weighted using the volume of the paired
+    storage node.
+
+    Parameters
+    ----------
+    storages: iterable of storage nodes.
```
Comment (jetuk) [inline pywr/parameters/_control_curves.pyx:563]: Should probably say these need to be the same length.

File: pywr/parameters/_control_curves.pyx [hunk 2]
```diff
@@ -548,3 +549,78 @@ cdef class ControlCurveParameter(BaseControlCurveParameter):
...
+        for (storage, profile) in zip(self.storages, self.profiles):
+            max_volume = storage.max_volume
+            if isinstance(max_volume, Parameter):
+                if max_volume.is_constant():
+                    max_volume = max_volume.get_constant_value()
+                else:
+                    raise ValueError("Storages with max_volume set to non-constant parameters cannot be used in WeightedAverageProfileParameter")
+            total_volume += max_volume
+
+            if isinstance(profile, ConstantParameter):
```
Comment (jetuk) [inline pywr/parameters/_control_curves.pyx:593]: Could this use `.is_constant()` to be a bit more generic?

File: pywr/parameters/_control_curves.pyx [hunk 3]
```diff
@@ -548,3 +549,78 @@ cdef class ControlCurveParameter(BaseControlCurveParameter):
         return self._upper_bounds
 
 ControlCurveParameter.register()
+
+
+cdef class WeightedAverageProfileParameter(Parameter):
+    """Generates a weighted average daily profile from pairs of storage nodes and
+    profile parameters.
+
+    The contribution on each profile is weighted using the volume of the paired
```
Comment (jetuk) [inline pywr/parameters/_control_curves.pyx:558]: Should mention that this parameter is efficient by calculating the profile during `.reset()`, but this limits the types of parameters that can be used.

File: tests/test_control_curves.py [hunk 4]
```diff
@@ -516,3 +522,117 @@ def test_json(self, simple_storage_model):
...
+
+        profiles = [curve0, curve1]
+
+        agg_curves = WeightedAverageProfileParameter(m, storages, profiles)
+
+        m.setup()
+
+        assert_array_almost_equal(agg_curves.get_daily_values(), np.full(366, 0.5))
+
+    def test_varying_curve(self, three_storage_model):
```
Comment (jetuk) [inline tests/test_control_curves.py:607]: Could this test the interp_day argument on the `MonthlyProfileParameter`?

**Review by jetuk — CHANGES_REQUESTED** [review]
Looks good from a technical perspective, but some comments on documentation and location of the new functions.

File: pywr/parameters/_control_curves.pyx [hunk 5]
```diff
@@ -548,3 +549,78 @@ cdef class ControlCurveParameter(BaseControlCurveParameter):
...
+    storage node.
+
+    Parameters
+    ----------
+    storages: iterable of storage nodes.
+    profiles: iterable of profile parameters. These can be either a
```
[the remark this answers is not in the document]
Comment (Batch21) [reply pywr/parameters/_control_curves.pyx:564]: Mainly for efficiency. For my current use case, the lists of storages/profiles is quite long for some of the instances, so there seems quite a bit on benefit in only calculating the curve once.

File: tests/test_control_curves.py (hunk 4 above)
Comment (Batch21) [reply tests/test_control_curves.py:607]: done

File: pywr/parameters/_control_curves.pyx [hunk 6]
```diff
@@ -548,3 +549,78 @@ cdef class ControlCurveParameter(BaseControlCurveParameter):
...
+                    total_absolute_profile += np.array(daily_values)
+            else:
+                raise ValueError("Control curve should be a ContantParameter, MonthlyProfileParameter, \
+                                  or DailyProfileParameter")
+        self.daily_values = total_absolute_profile / total_volume
+
+    cpdef double value(self, Timestep ts, ScenarioIndex scenario_index) except? -1:
+        return self.daily_values[ts.dayofyear_index]
+
+    cpdef double[:] get_daily_values(self):
```
[the remark this answers is not in the document]
Comment (Batch21) [reply pywr/parameters/_control_curves.pyx:613]: could do, may be in a separate PR? It might be tricky from some, such as the `WeeklyProfileParameter`

**Review by jetuk — APPROVED** [review]

**Pull request merged**